numpy/random/mtrand/mtrand.pyx

```diff
@@ -4901,10 +4901,13 @@ cdef class RandomState:
         """
         if isinstance(x, (int, long, np.integer)):
             arr = np.arange(x)
-        else:
-            arr = np.array(x)
-        self.shuffle(arr)
-        return arr
+            self.shuffle(arr)
+            return arr
+
+        arr = np.array(x)
+        idx = np.arange(arr.shape[0])
```

Comment:
If you use `arange(..., dtype=np.intp)` then you'll hit the optimized case in `shuffle`

Reply:
Done.